src/everest/everest_storage.py

```diff
@@ -0,0 +1,897 @@
+        to the user.
+        """
+        scaled_cols = [c for c in df.columns if c.lower().startswith("scaled")]
+        if len(scaled_cols) > 0:
+            raise ValueError("Scaled columns should not be stored into Everest storage")
+
+        # Keys are ROPT column keys
+        # values are corresponding column keys we present to the user
+        renames = {
+            "objective": "objective_name",
+            "weighted_objective": "total_objective_value",
+            "variable": "control_name",
+            "variables": "control_value",
+            "objectives": "objective_value",
+            "constraints": "constraint_value",
+            "nonlinear_constraint": "constraint_name",
+            "perturbed_variables": "perturbed_control_value",
+            "perturbed_objectives": "perturbed_objective_value",
+            "perturbed_constraints": "perturbed_constraint_value",
+            "evaluation_ids": "simulation_id",
+        }
+        return df.rename({k: v for k, v in renames.items() if k in df.columns})
+
+    @staticmethod
+    def _enforce_dtypes(df: polars.DataFrame) -> polars.DataFrame:
+        dtypes = {
+            "batch_id": polars.UInt32,
+            "perturbation": polars.UInt32,
+            "realization": polars.UInt32,
+            # -1 is used as a value in simulator cache.
+            # thus we need signed, otherwise we could do unsigned
+            "simulation_id": polars.Int32,
+            "perturbed_evaluation_ids": polars.Int32,
+            "objective_name": polars.String,
+            "control_name": polars.String,
+            "constraint_name": polars.String,
+            "total_objective_value": polars.Float64,
+            "control_value": polars.Float64,
+            "objective_value": polars.Float64,
+            "constraint_value": polars.Float64,
+            "perturbed_control_value": polars.Float64,
+            "perturbed_objective_value": polars.Float64,
+            "perturbed_constraint_value": polars.Float64,
+        }
+
+        existing_cols = set(df.columns)
+        unaccounted_cols = existing_cols - set(dtypes)
+        if len(unaccounted_cols) > 0:
+            raise KeyError(
+                f"Expected all keys to have a specified dtype, found {unaccounted_cols}"
+            )
+
+        df = df.cast(
+            {
+                colname: dtype
+                for colname, dtype in dtypes.items()
+                if colname in df.columns
+            }
+        )
+
+        return df
+
+    def write_to_output_dir(self) -> None:
+        exp = _OptimizerOnlyExperiment(self._output_dir)
+
+        # csv writing mostly for dev/debugging/quick inspection
+        self.data.write_to_experiment(exp)
+
+    def read_from_output_dir(self) -> None:
+        exp = _OptimizerOnlyExperiment(self._output_dir)
+        self.data.read_from_experiment(exp)
+
+    def observe_optimizer(
+        self,
+        optimizer: BasicOptimizer,
+        merit_file: Path,
```

Comment:
I would not pass the merit file path as an argument. As it is, this is an implementation detail to get merit handling to work with one specific dakota algorithm. I would hard-code somewhere in the merit-handling code, and get rid of the `self._merit` attribute. We need to remove this merit handling approach in the near future anywhen when we do better handling of constraint violations.